File: ios-native/Sources/PolarisStudyKit/WritingChecker.swift
import Foundation

/// Offline, deterministic writing checker (grammar / style / clarity) — no AI.
///
/// This is the on-device, works-offline tier of the writing assistant, mirroring
/// `packages/writing/rules.py`. The AI coach ("Polly") is an **online-only** feature: call its
/// API only when the device has connectivity; this checker always works.
public struct WritingIssue: Sendable, Equatable {
    public let range: Range<String.Index>
    public let type: String        // grammar | style | clarity | punctuation
    public let message: String
    public let suggestion: String?
    public let matched: String
}

public enum WritingChecker {

    private static let wordiness: [String: String] = [
        "in order to": "to", "due to the fact that": "because", "in the event that": "if",
        "at this point in time": "now", "a large number of": "many", "a majority of": "most",
        "for the purpose of": "to", "has the ability to": "can", "with regard to": "about",
        "each and every": "every", "in spite of the fact that": "although",
    ]
    private static let weasel: Set<String> = [
        "very", "really", "quite", "rather", "somewhat", "fairly", "actually", "basically",
        "clearly", "obviously", "simply", "just", "totally", "extremely",
    ]
    private static let confusables: [String: String] = [
        "its": "it's", "it's": "its", "their": "there/they're", "there": "their/they're",
        "your": "you're", "you're": "your", "affect": "effect", "effect": "affect",
        "then": "than", "than": "then", "loose": "lose",
    ]

    /// Analyze `text` and return located issues (grammar/style/clarity/punctuation).
    public static func check(_ text: String) -> [WritingIssue] {
        var issues: [WritingIssue] = []
        let lower = text.lowercased()

        // Wordy phrases.
        for (phrase, repl) in wordiness {
            for range in ranges(of: phrase, in: lower, source: text) {
                issues.append(.init(range: range, type: "clarity", message: "Wordy — tighten this",
                                    suggestion: repl, matched: String(text[range])))
            }
        }

        // Per-word: weasel words + confusables + repeated words.
        let words = tokenize(text)
        for (i, tok) in words.enumerated() {
            let w = tok.text.lowercased()
            if weasel.contains(w) {
                issues.append(.init(range: tok.range, type: "style",
                                    message: "Weak intensifier “\(tok.text)” — cut or strengthen",
                                    suggestion: "(delete)", matched: tok.text))
            } else if let alt = confusables[w] {
                issues.append(.init(range: tok.range, type: "grammar",
                                    message: "Commonly confused: check “\(tok.text)” vs “\(alt)”",
                                    suggestion: nil, matched: tok.text))
            }
            if i > 0, words[i - 1].text.lowercased() == w {  // repeated adjacent word
                let r = words[i - 1].range.lowerBound..<tok.range.upperBound
                issues.append(.init(range: r, type: "grammar", message: "Repeated word “\(tok.text)”",
                                    suggestion: tok.text, matched: String(text[r])))
            }
        }

        // Passive voice.
        passive(text, into: &issues)
        // Double spaces.
        for range in regexRanges("  +", in: text) {
            issues.append(.init(range: range, type: "punctuation", message: "Extra space",
                                suggestion: " ", matched: String(text[range])))
        }
        // Long sentences.
        for sentence in sentenceRanges(text) {
            let n = tokenize(String(text[sentence])).count
            if n > 30 {
                issues.append(.init(range: sentence, type: "clarity",
                                    message: "Long sentence (\(n) words) — consider splitting",
                                    suggestion: nil, matched: String(text[sentence]).trimmed(40)))
            }
        }

        return issues.sorted { $0.range.lowerBound < $1.range.lowerBound }
    }

    /// A 0–100 quality score from the issue mix (matches the Python scoring).
    public static func score(_ text: String, issues: [WritingIssue]) -> Int {
        let words = max(1, tokenize(text).count)
        let weights = ["grammar": 4, "clarity": 3, "style": 2, "punctuation": 1]
        let penalty = issues.reduce(0.0) { $0 + Double(weights[$1.type] ?? 2) }
        let density = penalty / (Double(words) / 100 + 1)
        return max(0, min(100, Int((100 - density).rounded())))
    }

    // MARK: - Helpers

    private struct Token { let text: String; let range: Range<String.Index> }

    private static func tokenize(_ text: String) -> [Token] {
        var tokens: [Token] = []
        var start: String.Index?
        for i in text.indices {
            if text[i].isLetter || text[i] == "'" {
                if start == nil { start = i }
            } else if let s = start {
                tokens.append(Token(text: String(text[s..<i]), range: s..<i)); start = nil
            }
        }
        if let s = start { tokens.append(Token(text: String(text[s...]), range: s..<text.endIndex)) }
        return tokens
    }

    private static func ranges(of phrase: String, in lower: String,
                               source: String) -> [Range<String.Index>] {
        var out: [Range<String.Index>] = []
        var search = lower.startIndex
        while let r = lower.range(of: phrase, range: search..<lower.endIndex) {
            // Map the lowercased offsets back onto the source (same length).
            let lo = source.index(source.startIndex, offsetBy: lower.distance(from: lower.startIndex, to: r.lowerBound))
            let hi = source.index(source.startIndex, offsetBy: lower.distance(from: lower.startIndex, to: r.upperBound))
            out.append(lo..<hi)
            search = r.upperBound
        }
        return out
    }

    private static func regexRanges(_ pattern: String, in text: String) -> [Range<String.Index>] {
        guard let re = try? NSRegularExpression(pattern: pattern) else { return [] }
        let ns = text as NSString
        return re.matches(in: text, range: NSRange(location: 0, length: ns.length)).compactMap {
            Range($0.range, in: text)
        }
    }

    private static func passive(_ text: String, into issues: inout [WritingIssue]) {
        let pattern = "\\b(is|are|was|were|be|been|being|get|got)\\s+\\w+ed\\b"
        for range in regexRanges(pattern, in: text) {
            issues.append(.init(range: range, type: "clarity",
                                message: "Passive voice — prefer an active subject",
                                suggestion: nil, matched: String(text[range])))
        }
    }

    private static func sentenceRanges(_ text: String) -> [Range<String.Index>] {
        var out: [Range<String.Index>] = []
        var start = text.startIndex
        for i in text.indices where ".!?".contains(text[i]) {
            let next = text.index(after: i)
            out.append(start..<next); start = next
        }
        if start < text.endIndex { out.append(start..<text.endIndex) }
        return out
    }
}

private extension String {
    func trimmed(_ n: Int) -> String {
        let t = trimmingCharacters(in: .whitespacesAndNewlines)
        return t.count <= n ? t : String(t.prefix(n)) + "…"
    }
}

File: android-native/polaris-study-kit/src/main/java/com/polaris/studykit/WritingChecker.kt
package com.polaris.studykit

import kotlin.math.roundToInt

/**
 * Offline, deterministic writing checker (grammar / style / clarity) — no AI.
 *
 * The on-device, works-offline tier of the writing assistant, mirroring
 * `packages/writing/rules.py`. The AI coach ("Polly") is an **online-only** feature — call its
 * API only when the device has connectivity (see [PolarisStudy] docs); this checker always works.
 */
data class WritingIssue(
    val start: Int,
    val end: Int,
    val type: String,          // grammar | style | clarity | punctuation
    val message: String,
    val suggestion: String? = null,
    val matched: String = "",
)

object WritingChecker {

    private val wordiness = mapOf(
        "in order to" to "to", "due to the fact that" to "because", "in the event that" to "if",
        "at this point in time" to "now", "a large number of" to "many", "a majority of" to "most",
        "for the purpose of" to "to", "has the ability to" to "can", "with regard to" to "about",
        "each and every" to "every", "in spite of the fact that" to "although",
    )
    private val weasel = setOf(
        "very", "really", "quite", "rather", "somewhat", "fairly", "actually", "basically",
        "clearly", "obviously", "simply", "just", "totally", "extremely",
    )
    private val confusables = mapOf(
        "its" to "it's", "it's" to "its", "their" to "there/they're", "there" to "their/they're",
        "your" to "you're", "you're" to "your", "affect" to "effect", "effect" to "affect",
        "then" to "than", "than" to "then", "loose" to "lose",
    )
    private val passive = Regex("\\b(is|are|was|were|be|been|being|get|got)\\s+\\w+ed\\b", RegexOption.IGNORE_CASE)
    private val wordRe = Regex("[A-Za-z']+")

    /** Analyze [text] and return located issues, sorted by position. */
    fun check(text: String): List<WritingIssue> {
        val issues = mutableListOf<WritingIssue>()
        val lower = text.lowercase()

        for ((phrase, repl) in wordiness) {
            var i = lower.indexOf(phrase)
            while (i >= 0) {
                issues.add(WritingIssue(i, i + phrase.length, "clarity", "Wordy — tighten this",
                    repl, text.substring(i, i + phrase.length)))
                i = lower.indexOf(phrase, i + phrase.length)
            }
        }

        val tokens = wordRe.findAll(text).toList()
        tokens.forEachIndexed { idx, m ->
            val w = m.value.lowercase()
            if (w in weasel) {
                issues.add(WritingIssue(m.range.first, m.range.last + 1, "style",
                    "Weak intensifier “${m.value}” — cut or strengthen", "(delete)", m.value))
            } else if (w in confusables) {
                issues.add(WritingIssue(m.range.first, m.range.last + 1, "grammar",
                    "Commonly confused: check “${m.value}” vs “${confusables[w]}”", null, m.value))
            }
            if (idx > 0 && tokens[idx - 1].value.lowercase() == w) {
                val s = tokens[idx - 1].range.first
                val e = m.range.last + 1
                issues.add(WritingIssue(s, e, "grammar", "Repeated word “${m.value}”",
                    m.value, text.substring(s, e)))
            }
        }

        for (m in passive.findAll(text)) {
            issues.add(WritingIssue(m.range.first, m.range.last + 1, "clarity",
                "Passive voice — prefer an active subject", null, m.value))
        }
        for (m in Regex("  +").findAll(text)) {
            issues.add(WritingIssue(m.range.first, m.range.last + 1, "punctuation", "Extra space", " ", m.value))
        }
        for (s in sentences(text)) {
            val n = wordRe.findAll(s.substring).count()
            if (n > 30) {
                issues.add(WritingIssue(s.start, s.end, "clarity",
                    "Long sentence ($n words) — consider splitting", null,
                    s.substring.trim().let { if (it.length <= 40) it else it.take(40) + "…" }))
            }
        }

        return issues.sortedBy { it.start }
    }

    /** A 0–100 quality score from the issue mix (matches the Python scoring). */
    fun score(text: String, issues: List<WritingIssue>): Int {
        val words = wordRe.findAll(text).count().coerceAtLeast(1)
        val weights = mapOf("grammar" to 4, "clarity" to 3, "style" to 2, "punctuation" to 1)
        val penalty = issues.sumOf { (weights[it.type] ?: 2).toDouble() }
        val density = penalty / (words / 100.0 + 1)
        return (100 - density).roundToInt().coerceIn(0, 100)
    }

    private data class Span(val start: Int, val end: Int, val substring: String)

    private fun sentences(text: String): List<Span> {
        val out = mutableListOf<Span>()
        var start = 0
        for (i in text.indices) {
            if (text[i] in ".!?") {
                out.add(Span(start, i + 1, text.substring(start, i + 1))); start = i + 1
            }
        }
        if (start < text.length) out.add(Span(start, text.length, text.substring(start)))
        return out
    }
}
